pkg/bloomgateway/client.go

```diff
@@ -245,47 +215,180 @@ func (c *GatewayClient) doForAddrs(addrs []string, fn func(logproto.BloomGateway
 	return err
 }
 
-// serverAddrsForFingerprints returns a slices of server address slices for
-// each fingerprint of given fingerprints.
-// The indexes of the returned slices correspond to each other.
-// Returns an error in case the bloom gateway ring could not get the
-// corresponding replica set for a given fingerprint.
-// Warning: This function becomes inefficient when the number of fingerprints is very large.
-func (c *GatewayClient) serverAddrsForFingerprints(tenantID string, groups []*logproto.GroupedChunkRefs) ([]uint64, [][]string, error) {
-	subRing := GetShuffleShardingSubring(c.ring, tenantID, c.limits)
+func (c *GatewayClient) groupFingerprintsByServer(groups []*logproto.GroupedChunkRefs, subRing ring.ReadRing, instances []ring.InstanceDesc) ([]instanceWithFingerprints, error) {
```

Comment:
Do you think it makes sense to move this whole logic to somewhere near `pkg/util/ring/sharding.go` so it can be shared?

Reply:
in other words, can we make use of it for bloom-compactor?